### ISSUE TITLE
Broker runs into "disk full" issue when not being used

### ISSUE THREAD

**Issue by ctron** [EnMasseProject/enmasse#1440]
Configuring an address of type `pooled-queue` and not publishing messages to this address causes the disk to run full after 6 days:

PVC:
~~~
Requested Capacity:    2 GiB 
~~~

Disk usage on the container:
~~~
sh-4.2$ df .
Filesystem                                         1K-blocks    Used Available Use% Mounted on
10.16.154.241:vol_4a6ea60dd7917536abb3197b16b8c03b   2086912 2086912         0 100% /run/artemis
sh-4.2$ du -h -d 1
12K     ./bin
26K     ./etc
606K    ./log
1.8G    ./tmp
4.0K    ./lib
23M     ./data
4.0K    ./lock
1.8G    .
sh-4.2$ command terminated with non-zero exit code: Error executing in Docker Container: 137
~~~

Using EnMasse 0.21.0 on OpenShift OCP 3.9.

**Comment by k-wall** [comment]
@ctron - thanks I'll take a look

**Comment by ctron** [comment]
@k-wall Sorry, the instances are gone now. I can try to reproduce this, maybe just create a standard setup and let it run for a while. But note that you won't run into "disk full" if you are using NFS or local disk, as there are no quotas enforced.

**Comment by k-wall** [comment]
@ctron I tried to reproduce myself but was unsuccessful.  The storage utilisation by the ```tmp``` was completely flat.   The only way I could force a growth was to force a JVM crash.  I was speculating that if you had encountered a situation where the Broker was periodically suffering such a crash for some reason, that my account for the grow you saw leading to the disk full.

**Comment by ctron** [comment]
> I was speculating that if you had encountered a situation where the Broker was periodically suffering such a crash for some reason, that my account for the grow you saw leading to the disk full.

Which would still be case to fix, right?

**Comment by k-wall** [comment]
Currently there is configuration to point Java's tmp location ```java.io.tmpdir``` to a location within the persistent volume.  I think this is questionable choice.   Even if Artemis or its JVM crashes for an unexpected reason, we'd still want any temporary files to go without manual intervention.   Perhaps this should be repointed to the standard ```tmp`` location of the pod.

I'd still like to understamnd the underlying reason for the build-up of files, so if @ctron, you are able to reproduce this issue, that would be appreciated.

**Comment by ctron** [comment]
> I'd still like to understamnd the underlying reason for the build-up of files, so if @ctron, you are able to reproduce this issue, that would be appreciated.

Right now I have trouble to. I don't have any machine which I can let run for 6 days straight in a similar environment.

**Comment by k-wall** [comment]
It is the Artemis start script that overrides ```java.io.tmpdir``` to a value beneath on the ```$ARTEMIS_INSTANCE``` directory.   Artemis provides no way to separate the temporary location from the instance directory parent.      I had hoped that one could override the ```java.io.tmpdir``` within an ```_JAVA_OPTIONS``` environment variable (as these always trump Java command line arguments), but unfortunately ```org.apache.activemq.artemis.boot.Artemis#execute()``` actually overwrites ```java.io.tmpdir``` a second time, so any changes from outside are lost.

I think there needs to be a Artemis change.  In a world of containers, I think separating the temporary files from persistent data will be a common request.  Another possibility would be an Artemis change to allow ```org.eclipse.jetty.webapp.WebAppContext#setPersistTempDirectory``` to be set true so that Jetty's temporary location is stable between restarts.   However, this might complicate upgrade procedures.

Without an Artemis change, I see only one workaround.   It is possible to have EnMasses's ```launch.sh``` create a symlink from the $ARTEMIS_INSTANCE/tmp to a temporary location created with ```$(tmpfile -d)```.  This is a hack.  It means the persistent volume a symlink that sometimes dangles.

**Comment by k-wall** [comment]
Raised upstream issue against Artemis: https://issues.apache.org/jira/browse/ARTEMIS-2026

**Comment by k-wall** [comment]
Closing, Artemis issue ARTEMIS-2026.

**Comment by vhdirk** [comment]
Upstream fixed (https://issues.apache.org/jira/browse/ARTEMIS-2026) to be released in Artemis 2.17.0, apparently.
